### ISSUE TITLE
Internal Server Error when filtering on self-referencing m2m relation

### ISSUE THREAD

**Issue by Tapematch** [directus/directus#5747]
Hey guys,
thank you very much for this amazing piece of software, you are doing incredible work.

Unfortunately, the API is returning a 500 internal server error when I try to filter items based on a self-referencing m2m-relationship.

Let's say I want to serve multiple _blogs_ with one directus project. One blog can have multiple _articles_ (o2m) and has multiple writers (o2m to _directus_users_). Now, every blog can have partnerships with many blogs (self-referencing m2m relationship).

![screenshot](https://user-images.githubusercontent.com/17563492/118635707-2d8b8480-b7d4-11eb-889d-2fa6f05b7aab.png)

Now I want to give a user the right to read all articles of all partner blogs of their own blog. I would set a filter on reading articles like this:

`{
    "blog": {
        "partner_blogs": {
            "blog_related_id": {
                "users": {
                    "_eq": "$CURRENT_USER"
                }
            }
        }
    }
}`

An API request on the article-items with this filter returns the following error:

`{
    "errors": [
        {
            "message": "select \"article\".\"id\", \"article\".\"title\", \"article\".\"blog\" from \"article\" left join \"blog\" as \"telfk\" on \"article\".\"blog\" = \"telfk\".\"id\" left join \"blog_blog\" as \"givpw\" on \"telfk\".\"id\" = \"givpw\".\"blog_id\" where \"blog\".\"users\" = $1 order by \"article\".\"id\" asc limit $2 - ungültiger Verweis auf FROM-Klausel-Eintrag für Tabelle »blog«",
            "extensions": {
                "code": "INTERNAL_SERVER_ERROR"
            }
        }
    ]
}`

I am using the latest Directus RC(69) and PostgreSQL 13.2.
I created a minimal collection structure and attached an sql dump. Thanks for your help!

[dump.zip](https://github.com/directus/directus/files/6500615/dump.zip)

**Comment by rijkvanzanten** [comment]
Just to confirm my limited german, this `ungültiger Verweis auf FROM-Klausel-Eintrag für Tabelle »blog«` roughly means "Invalid reference in FROM statement for the "blog" table" right?

**Comment by Tapematch** [comment]
Yes, the message is `invalid reference to FROM-clause entry for table \"blog\"` when I switch to english.

**Comment by rijkvanzanten** [comment]
It doesn't matter for the bug in related blogs, but just to confirm:

> and has multiple writers (o2m to directus_users)

Can a single writer be in multiple blogs? (eg should this be a m2m instead of a o2m)

**Comment by Tapematch** [comment]
Yes that would also make sense, but I tried to keep it as simple as possible for the demonstration :smiley:

**Comment by rijkvanzanten** [comment]
@Tapematch Thanks for the dump, really helpful 🙂 

In your filter, `users` is a one-to-many item, which itself is not a "column" with a "value". It's a virtual reference to a set of related items, but isn't a column that holds a value itself. You can't use _eq on this non-existing column. That being said, you can filter on the references within. Changing your filter from 

```json
{
  "blog": {
    "partner_blogs": {
      "blog_related_id": {
        "users": {
          "_eq": "$CURRENT_USER"
        }
      }
    }
  }
}
```

to

```json
{
  "blog": {
    "partner_blogs": {
      "blog_related_id": {
        "users": {
          "id": {
            "_eq": "$CURRENT_USER"
          }
        }
      }
    }
  }
}
```

should work as expected 👍🏻

**Comment by Tapematch** [comment]
Hey, I'm not really sure if I am doing something wrong, but your suggested filter doesn't seem to work for me either.
I'm getting a slightly different error message:

`error: select "article"."id", "article"."title", "article"."blog" from "article" left join "blog" as "mgbrk" on "article"."blog" = "mgbrk"."id" left join "blog_blog" as "cinpp" on "mgbrk"."id" = "cinpp"."blog_id" where ("directus_users"."id" = $1) order by "article"."id" asc limit $2 - missing FROM-clause entry for table "directus_users"`

However, I tried to do some debugging myself and I got to this line:
https://github.com/directus/directus/blob/48cdf6e0835b6e077cb0641be30d49483c611439/api/src/utils/apply-query.ts#L123
If I comment that out and do the _followRelation_ until the end of the chain it seems to work as expected. I had another issue where I have a filter on a self-referencing m2o field, that was also solved with this change.

It seems the _subQuery_ parameter isn't used anywhere, and I can't really tell if the check for a m2o relation is needed there, because the _pathParts_ will not be followed through the end if we check it.

Maybe you have another suggestion for making my filter query work?
Thanks!

**Comment by rijkvanzanten** [comment]
Ah yeah you're correct, my bad! I was testing it on the latest `main`, which includes two big changes to relational fetching and checking (https://github.com/directus/directus/pull/5615 and https://github.com/directus/directus/pull/5855) which seem to have fixed this problem as well 🙂 

> It seems the subQuery parameter isn't used anywhere, and I can't really tell if the check for a m2o relation is needed there, because the pathParts will not be followed through the end if we check it.

This is false. Subquery is used when fetching a nested o2m data set, to prevent duplicates on the parent level 🙂 (https://github.com/directus/directus/blob/bd6cab8989ea8a8af3b0848c51ee9b3fae289ab3/api/src/utils/apply-query.ts#L283)

**Comment by rijkvanzanten** [comment]
From @Tapematch (https://github.com/directus/directus/discussions/6202)

---

Hey guys,
I still need help with my issue #5747.
I tried it with the newest RC75, but am still getting the same errors. I created my collections from scratch to rule out any migration errors.

Here a small summary:
1. Create a self-referencing m2m relation
2. Try filtering items based on that relation

This is the filter I'm trying to use on the `article` collection:
```
{
  "blog": {
    "partner_blogs": {
      "related_blog_id": {
        "users": {
          "id": {
            "_eq": "$CURRENT_USER"
          }
        }
      }
    }
  }
}
```

And this the error it generates:
```
error: select "article"."id", "article"."title", "article"."blog" from "article" left join "blog" as "ciakr" on "article"."blog" = "ciakr"."id" left join "blog_blog" as "khjic" on "ciakr"."id" = "khjic"."blog_id" where "directus_users"."id" = $1 order by "article"."id" asc limit $2 - missing FROM-clause entry for table "directus_users"
```

I also attached a new database dump. Maybe someone here can point out my errors. I am grateful for any help. Thanks!

[dump.zip](https://github.com/directus/directus/files/6636660/dump.zip)
_Admin user is admin@example.com:12345678. You can use the token 12345678 for requesting._

**Comment by philleepflorence** [comment]
The following SQL statement lacks a `JOIN` to `directus_users` but is referenced in `WHERE`:

```sql
SELECT "article"."id", "article"."title", "article"."blog" 
FROM "article" 
LEFT JOIN "blog" AS "ciakr" ON "article"."blog" = "ciakr"."id" 
LEFT JOIN "blog_blog" AS "khjic" ON "ciakr"."id" = "khjic"."blog_id" 
WHERE "directus_users"."id" = $1 
ORDER BY "article"."id" ASC LIMIT $2;
```
Only `article`, `blog`, and `blog_blog` are referenced.

It looks like there is no relationship between `blog` and `directus_users`, not sure if it is user error or a code bug.

**Comment by techpet** [comment]
Any news on this? Similar issue here using directus v9.0.0-rc.83 with Postgres v13:

![diagram3](https://user-images.githubusercontent.com/1122529/125215245-87379900-e2c3-11eb-9012-24c9a6fea000.png)

Trying to filter units belonging to a branch based on a user's company participating to an assignment relationship (branch_company table) with the units' branch.

A branch has a M2M Assignment relation with a company:

![branch_assignments](https://user-images.githubusercontent.com/1122529/125215369-00cf8700-e2c4-11eb-9476-4deaf8341f67.png)

And a company has a M2M Assignment relation with directus_users:

![company_users](https://user-images.githubusercontent.com/1122529/125215412-1a70ce80-e2c4-11eb-9645-4cbefde812f6.png)

The query used for unit permissions is:

```
{
    "branch": {
        "assignments": {
            "company_id": {
                "users": {
                    "directus_users_id": {
                        "_eq": "$CURRENT_USER"
                    }
                }
            }
        }
    }
} 
```

and the error message:
``` 
error: select "unit"."id" from "unit" left join "branch" as "uyqcx" on "unit"."branch" = "uyqcx"."id" left join "branch_company" as "fvede" on "uyqcx"."id" = "fvede"."branch_id" where ("company_directus_users"."directus_users_id" = $1) order by "unit"."id" asc limit $2 - missing FROM-clause entry for table "company_directus_users"
```

**Comment by massimoliani** [comment]
Still an open issue, I check all related issues, but can't fix.
It look like starting from a M2O relation it stop before the last join.

**Comment by massimoliani** [comment]
I still have problems with filter.
I found that applyFilter function in apply-query.js work well starting with a not m2o relation in the filter.
First it add one single join then all other joins are added in the where clause.
When the filter has a m2o relation as first relation the addwhere clouse don't check if the filter continue and simply apply the where clause without following relations.

**Comment by ksoltanidev** [comment]
https://github.com/directus/directus/blob/1b3504b0b53565d6767566bc00116a456e0f0a82/api/src/utils/apply-query.ts#L202

Changed this line to this : 

`if (relationType === 'm2o' || relationType === 'o2m' || subQuery === true) {`

I tested it on my setup with issue #8493 and it seems to work. Directus adds the missing Joins to the request.
I hope I did not broke anything else by doing this. I guess this will need a bit of testing.

**Comment by ksoltanidev** [comment]
I just tested and it breaks other things.
I added this Read permission filter on the "pythagore_affaires" collection (db in #8493) and tried to get the data as a user (GET directusLink/items/pythagore_affaires)

```json
{
	"affairs": {
		"user_access": {
			"directus_users_id": {
				"_eq": "$CURRENT_USER"
			}
		}
	}
}
```

It's supposed, for a certain user, to allow access to only "pythagore_affaires" linked to a "affairs" which is linked to the "directus_users". With the "fix" I suggested, SQL request is broken.

Although, without it, it does not work properly either.I should get:

```json
{
    "data": [
        {
            "numero_affaire": "94",
            "affairs": [
                94
            ]
        }
    ]
}
```

But I get all "pythagore_affaires", even thoses where no "affairs" linked to the current user is found: 

```json
{
    "data": [
        {
            "numero_affaire": "94",
            "affairs": [
                94
            ]
        },
        {
            "numero_affaire": "95",
            "affairs": []
        }
    ]
}
```

**Comment by alex4o** [comment]
I think I have the same problem, with m2m relations, but I am not sure. I have `directus_users` and the table that has 2 columns that link to it.

It generates this: 
```sql
SELECT "invoices"."id"
FROM "invoices"
LEFT JOIN "directus_users" AS "pckds" ON "invoices"."uploader" = "pckds"."id"
LEFT JOIN "junction_accountants" AS "bgecn" ON "pckds"."id" = "bgecn"."user"
WHERE ("directus_users"."id" = $1)
```

When I want something akin to:

```sql

SELECT *
FROM "invoices"
LEFT JOIN "junction_accountants" ON "invoices"."uploader" = "junction_accountants"."user"
WHERE "junction_accountants"."accountant" = $1;
```

**Comment by massimoliani** [comment]
I've seen the error happen when you start with a m2o anche then you continue with m2m.
It look like the first m2o stop the processing of following relationship.
I fix momentary the problem adding an hook before save that add an extra field for following relations so I can always filter on that relations. 
The hook grant you you never miss an update.

**Comment by jkarelins** [comment]
Hello everyone, 
Is it already fixed? I am using latest Ditrectus, and then I have found this issue, closed my issue as a duplicate: #10470.

But I am still curious, if there is a way to fix it?

**Comment by rijkvanzanten** [comment]
@jkarelins The issue is still open, indicating that it wasn't resolved yet. 

> But I am still curious, if there is a way to fix it?

There is, though nobody has provided a PR to fix it yet. Wanna try fixing it? 🙂

**Comment by jkarelins** [comment]
> @jkarelins The issue is still open, indicating that it wasn't resolved yet. 
> 
> > But I am still curious, if there is a way to fix it?
> 
> There is, though nobody has provided a PR to fix it yet. Wanna try fixing it? 🙂 

Will see on the weekend, what I can do.

**Comment by rijkvanzanten** [comment]
Fixed in #11159
